### PR TITLE
[Merged by Bors] - Tls authentication and encryption

### PR DESCRIPTION
**Pull request by maltesander**
## Description

closes https://github.com/stackabletech/zookeeper-operator/issues/466

Currently there is still a workaround (TLS can be used via the `clientPort` and not `secureClientPort` with port unification. This means there still can be plaintext connections without encryption or authentication; see https://issues.apache.org/jira/browse/ZOOKEEPER-4276)

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: deploy/config-spec/properties.yaml [hunk 1]
```diff
@@ -142,4 +144,258 @@ properties:
...
+        - "need"
+      defaultValues:
+        - value: "need"
+      recommendedValues:
+        - value: "need"
+      roles:
+        - name: "server"
+          required: false
+      asOfVersion: "3.5.7"
+      description: "Specifies options to authenticate ssl connections from clients."
```
Comment (sbernauer) [inline deploy/config-spec/properties.yaml:190]:
```suggestion
      description: "Specifies options to authenticate ssl connections from other quorum members."
```

File: rust/operator-binary/src/zk_controller.rs [hunk 2]
```diff
@@ -439,20 +519,126 @@ fn build_server_rolegroup_statefulset(
...
+            empty_dir: Some(EmptyDirVolumeSource {
+                medium: None,
+                size_limit: None,
+            }),
+            name: "rwconfig".to_string(),
+            ..Volume::default()
+        })
+        .add_volume(Volume {
+            name: "quorum-tls".to_string(),
+            csi: Some(CSIVolumeSource {
```
Comment (sbernauer) [inline rust/operator-binary/src/zk_controller.rs:575]: We do actually have a builder for that.
FYI: The csi-volume will be deprecated soon and will be replaced by ephemeral volumes by https://github.com/stackabletech/operator-rs/pull/396

File: rust/operator-binary/src/command.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            "expr $MYID_OFFSET + $(echo $POD_NAME | sed 's/.*-//') > {dir}/myid",
+            dir = STACKABLE_DATA_DIR
+        ),
+    ]);
+
+    args.join(" && ")
+}
+
+fn generate_password() -> String {
+    // taken from https://unix.stackexchange.com/questions/230673/how-to-generate-a-random-string
```
Comment (adwk67) [inline rust/operator-binary/src/command.rs:66]: I'd remove this link

Comment (adwk67) [inline rust/operator-binary/src/command.rs:66]:
```suggestion
```

File: rust/operator-binary/src/zk_controller.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator-binary/src/zk_controller.rs]: changed to builder

**Review by adwk67 — APPROVED** [review]
LGTM

File: rust/crd/src/lib.rs [hunk 4]
```diff
@@ -36,6 +47,67 @@ pub struct ZookeeperClusterSpec {
     pub version: Option<String>,
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub servers: Option<Role<ZookeeperConfig>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub config: Option<GlobalZookeeperConfig>,
+}
+
+#[derive(Clone, Default, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GlobalZookeeperConfig {
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:56]: Do you think it's worth it extracting that into a separate struct? I personally find it harder to read

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -103,6 +197,106 @@ impl Configuration for ZookeeperConfig {
...
+        );
+        result.insert(
+            Self::SSL_AUTH_PROVIDER_X509.to_string(),
+            Some("org.apache.zookeeper.server.auth.X509AuthenticationProvider".to_string()),
+        );
+        // The keystore and truststore passwords should not be in the configmap and are generated
+        // and written later via script in the init container
+        result.insert(
+            Self::SSL_QUORUM_KEY_STORE_LOCATION.to_string(),
+            Some(format!("{dir}/keystore.p12", dir = QUORUM_TLS_DIR)),
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:227]: You like to prefer extracting the format args, aren't you?

File: rust/crd/src/lib.rs (hunk 4 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:56]: No but the ADR stated something like:
```
config:
   quorum: xxx
   client: yyy
```

File: rust/crd/src/lib.rs (hunk 5 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:227]: yes!

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**